src/handlers.rs

```diff
@@ -65,3 +69,115 @@ pub async fn decode(req: web::Json<Request>) -> Result<impl Responder> {
     };
     Ok(web::Json(response))
 }
+
+async fn get_tx_logs(txn_hash: &DisplayBytes, network: &String) -> Result<Vec<TxLog>> {
+    let res = reqwest::get(format!(
+        "https://blockscout.com/{}/api?module=transaction&action=gettxinfo&txhash={}",
+        network, txn_hash
+    ))
+    .await
+    .map_err(|err| anyhow!(err))?;
+
+    if res.status() == StatusCode::NOT_FOUND {
+        return Err(Error::GetTxInfo("Wrong network".to_string()));
+    }
+
+    let res: Transaction = res.json().await.map_err(|err| anyhow!(err))?;
+
+    if res.status != "1" {
+        return Err(Error::GetTxInfo(res.message));
+    }
+
+    res.result
+        .map(|result| result.logs)
+        .ok_or_else(|| Error::Other(anyhow!("Missing input block")))
+}
+
+async fn get_contract_abi(network: &String, address: &DisplayBytes) -> Result<Contract> {
+    let res = reqwest::get(format!(
+        "https://blockscout.com/{}/api?module=contract&action=getabi&address={}",
+        network, address
+    ))
+    .await
+    .map_err(|err| anyhow!(err))?;
+
+    if res.status() == StatusCode::NOT_FOUND {
+        return Err(Error::GetTxInfo("Wrong network".to_string()));
+    }
+
+    let res: AbiResponse = res.json().await.map_err(|err| anyhow!(err))?;
+
+    if res.status == "0" {
+        return Err(Error::GetTxInfo(res.message));
+    }
+
+    let c: Contract = serde_json::from_str(
+        &res.result
+            .ok_or_else(|| anyhow!("Abi parsing failed".to_string()))?,
+    )
+    .map_err(|err| anyhow!(err))?;
+
+    Ok(c)
+}
+
+async fn get_log_event(
+    network: &String,
+    log: &TxLog,
+    abi_map: &mut HashMap<Vec<u8>, Contract>,
+) -> Result<Event> {
+    abi_map
+        .entry(log.address.to_vec())
+        .or_insert(get_contract_abi(network, &log.address).await?);
+
+    let abi = abi_map
+        .get(&log.address.to_vec())
+        .ok_or_else(|| anyhow!("Failed to get abi".to_string()))?;
+
+    for event in abi.events() {
+        if log.topics[0]
+            .clone()
```

Comment:
```suggestion
            .as_ref()
```

Nit: You may eliminate cloning here, by taking a reference for the value inside an Option